packages/modules/e3dc/device.py

```diff
@@ -0,0 +1,139 @@
+                           component_config)
+
+    def create_inverter_component(component_config: E3dcInverterSetup):
+        return E3dcInverter(device_config.id,
+                            component_config)
+
+    def update_components(components: Iterable[Union[E3dcBat, E3dcCounter, E3dcInverter]]):
+        with modbus.ModbusTcpClient_(device_config.configuration.address, 502) as client:
+            log.debug('reading: %s', device_config.configuration.address)
+            for component in components:
+                with SingleComponentUpdateContext(component.component_info):
+                    component.update(client)
+
+    return ConfigurableDevice(
+        device_config=device_config,
+        component_factory=ComponentFactoryByType(
+            bat=create_bat_component,
+            counter=create_counter_component,
+            inverter=create_inverter_component,
+        ),
+        component_updater=update_components
+    )
+
+
+def run_device_legacy(device_config: E3dc, component_config: Union[E3dcBatSetup, E3dcCounterSetup, E3dcBatSetup]):
+    device = create_device(device_config)
+    device.add_component(component_config)
+    log.debug("E3dc Configuration: %s, Component Configuration: %s", device_config, component_config)
+    device.update()
+
+
+def create_legacy_device_config(address: str,
+                                read_ext: int,
+                                num: int) -> None:
+    device_config = E3dc(configuration=E3dcConfiguration(address=address,
+                                                         read_ext=read_ext),
+                         id=num)
+    log.debug("Konfig: %s", device_config)
+    return device_config
+
+
+def read_legacy_counter(address1: str,
+                        address2: str,
+                        read_ext: int,
+                        pvinv: str,
+                        num: int):
+    component_config = E3dcCounterSetup(configuration=E3dcCounterConfiguration())
+    component_config.id = num
+    run_device_legacy(create_legacy_device_config(address1,
+                                                  read_ext,
+                                                  num), component_config)
+
+
+def read_legacy_bat(address1: str,
+                    address2: str, read_ext: int,
+                    pvinv: str,
+                    num: int) -> None:
+    # für openwbv19 können mit der bisherigen parametrisierung zwei ip_addressen
+    # ausgelesen werden
+    # ebenso wird bei Speicheraufruf beides (Speicher und PV ausgelesen
+    # in openwb v2.0 geht nur noch eine IP adresse und die Pv muss
+    # separate ausgelesen werden
+    addresses = [address for address in [address1, address2] if address != "none"]
+    log.debug('e3dc IP-Adresse1: ' + address1)
+    log.debug('e3dc IP-Adresse2: ' + address2)
+    log.debug('e3dc read_ext: ' + str(read_ext))
+    log.debug('e3dc pvinv: ' + pvinv)
+    log.debug('e3dc id: ' + str(num))
```

Comment:
Ob es nötig ist das auszugeben mag diskussionswürdig sein, denn der legacy-run-Server gibt auch schon alle Parameter aus (wenn auch als array und ggf. gekürzt). Wenn du es doch drin haben willst, dann nutze doch die Format-Funktion vom Loggingframework:
```suggestion
    log.debug("e3dc IP-Adresse1: %s", address1)
    log.debug("e3dc IP-Adresse2: %s", address2)
    log.debug("e3dc read_ext: %d", read_ext)
    log.debug("e3dc pvinv: %s", pvinv)
    log.debug("e3dc id: %d", num)
```

Reply:
geändert